### ISSUE TITLE
any.h dereferences a NULL pointer

### ISSUE THREAD

**Issue by KenYN** [mariusmuja/flann#346]
Looking at the code in [any.h/reset()](https://github.com/mariusmuja/flann/blob/master/src/cpp/flann/util/any.h#L261), I see:

    policy->static_delete(&object);

The [default implementation for `static_delete` is just](https://github.com/mariusmuja/flann/blob/06a49513138009d19a1f4e0ace67fbff13270c69/src/cpp/flann/util/any.h#L64):

    virtual void static_delete(void**) { }

But even with a default empty implementation like this, I think the official C++ standard would say that this is undefined behaviour when `object` is NULL. This problematic behaviour happens when one declared an uninitialised `any`, then assigns a new value. The `assign()` method calls `reset()` with `object` equal to NULL. A simple:

    if (object) policy->static_delete(&object);

Should avoid this problem.

**Comment by victl** [comment]
No, this is not a problem. Because the implementation of `any.h` uses the pointer itself as the storage for small data types (plain old datatypes).

If the underlying policy type is `small_any_policy`, then the pointer is not pointing any other address, but only hosting the content itself. So delete of the pointer is not needed.
